### PR TITLE
PLFM-5064 and PLFM-5065 manager

### PR DESCRIPTION
**Pull request by nickgros-sage**

### CODE REVIEW

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/doi/DoiManager.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+	 * @param objectId The ID of the object in Synapse
+	 * @param objectType The type of the object
+	 * @param versionNumber The version of the object. If null, refers to the most recent version.
+	 * @return The data transfer object for the DOI association.
+	 */
+	DoiAssociation getDoiAssociation(final Long userId, final String objectId, final ObjectType objectType, final Long versionNumber);
+
+	/**
+	 * Retrieves a DOI with all associated metadata.
+	 * Note that this method calls an external API, which may affect responsiveness.
```
Comment (john-hill) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/doi/DoiManager.java:35]: Idempotent

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/doi/DoiManagerImpl.java [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+		try {
+			if (!doiAssociationDao.getEtagForUpdate(dto.getObjectId(), dto.getObjectType(), dto.getObjectVersion()).equals(dto.getEtag())) {
+				throw new ConflictingUpdateException("eTags do not match.");
+			}
+			doiAssociationDao.updateDoiAssociation(dto);
+		} catch (NotFoundException e1) {
+			try {
+				doiAssociationDao.createDoiAssociation(dto); //retry on fail
+			} catch (Exception e2) {
+				throw new RecoverableMessageException(e2);
```
Comment (john-hill) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/doi/DoiManagerImpl.java:107]: This needs a comment as to why we are doing this.

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/doi/DoiManagerImpl.java [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+		String uri = generateDoiUri(dto.getObjectId(), dto.getObjectType(), dto.getObjectVersion());
+		String url = generateLocationRequestUrl(dto.getObjectId(), dto.getObjectType(), dto.getObjectVersion());
+		try {
+			dataciteClient.registerMetadata(dto, uri);
+			dataciteClient.registerDoi(uri, url);
+			if (dto.getStatus() == DataciteRegistrationStatus.REGISTERED) {
+				dataciteClient.deactivate(uri);
+			}
+		} catch (NotReadyException | ServiceUnavailableException e) {
+			throw new RecoverableMessageException(e);
```
Comment (john-hill) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/doi/DoiManagerImpl.java:120]: explain why this is recoverable.

File: services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/doi/DoiManagerImplTest.java [hunk 4]
```diff
@@ -0,0 +1,567 @@
...
+import org.sagebionetworks.repo.model.doi.v2.DoiTitle;
+import org.sagebionetworks.repo.web.NotFoundException;
+import org.sagebionetworks.repo.web.ServiceUnavailableException;
+import org.sagebionetworks.workers.util.aws.message.RecoverableMessageException;
+import org.springframework.dao.DuplicateKeyException;
+import org.springframework.test.context.ContextConfiguration;
+import org.springframework.test.util.ReflectionTestUtils;
+
+@RunWith(MockitoJUnitRunner.class)
+@ContextConfiguration(locations = { "classpath:test-context.xml" })
```
Comment (john-hill) [inline services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/doi/DoiManagerImplTest.java:47]: not needed for mock test

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/doi/DoiManagerImpl.java [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+		// Retrieve our record of the DOI/object association.
+		// Authorization is determined in the retrieval method
+		DoiAssociation association = getDoiAssociation(userId, objectId, objectType, versionNumber);
+
+		// Get the metadata from DataCite. If their API is down, this may fail with NotReadyException/ServiceUnavailableException
+		DataciteMetadata metadata = null;
+		try {
+			metadata = dataciteClient.get(association.getDoiUri());
+		} catch (NotReadyException | ServiceUnavailableException e) {
+			throw new ServiceUnavailableException(e);
```
Comment (john-hill) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/doi/DoiManagerImpl.java:51]: remove catch of ServiceUnavailableException (let it go through)

File: services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/doi/DoiManagerImplTest.java [hunk 6]
```diff
@@ -0,0 +1,567 @@
...
+		when(mockAuthorizationManager.canAccess(any(UserInfo.class), any(String.class), any(ObjectType.class), any(ACCESS_TYPE.class)))
+				.thenReturn(new AuthorizationStatus(false, "mock"));
+		doiManager.getDoi( 12345L, entityId, entityType, version);
+	}
+
+	@Test(expected = ServiceUnavailableException.class)
+	public void testGetDoiServiceUnavailableException() throws Exception {
+		DoiAssociation mockAssociation = setUpDto(false);
+		when(mockDoiDao.getDoiAssociation(entityId, entityType, version)).thenReturn(mockAssociation);
+		when(mockDataciteClient.get(any(String.class))).thenThrow(new ServiceUnavailableException());
```
Comment (john-hill) [inline services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/doi/DoiManagerImplTest.java:180]: Convert this to test the NotReadyException

File: services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/doi/DoiManagerImplTest.java [hunk 7]
```diff
@@ -0,0 +1,567 @@
...
+		assertEquals(mockResponse.getObjectId(), actualResponse.getObjectId());
+		assertEquals(mockResponse.getObjectType(), actualResponse.getObjectType());
+		assertEquals(mockResponse.getObjectVersion(), actualResponse.getObjectVersion());
+		assertEquals(mockPrefix + "/" + mockResponse.getObjectId() + "." + mockResponse.getObjectVersion(), actualResponse.getDoiUri());
+		assertEquals(doiManager.generateLocationRequestUrl(mockResponse.getObjectId(), mockResponse.getObjectType(), mockResponse.getObjectVersion()), actualResponse.getDoiUrl());
+	}
+
+	@Test(expected = UnauthorizedException.class)
+	public void testGetAssociationUnauthorized() {
+		when(mockAuthorizationManager.canAccess(any(UserInfo.class), any(String.class), any(ObjectType.class), any(ACCESS_TYPE.class)))
```
Comment (john-hill) [inline services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/doi/DoiManagerImplTest.java:117]: instead of any() target the real values.

File: services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/doi/DoiManagerImplTest.java [hunk 8]
```diff
@@ -0,0 +1,567 @@
...
+				.thenReturn(new AuthorizationStatus(false, "mock"));
+		// Call under test
+		doiManager.getDoiAssociation( 12345L, entityId, entityType, version);
+	}
+
+	@Test(expected = NotFoundException.class)
+	public void testGetAssociationNotFoundException() {
+		when(mockDoiDao.getDoiAssociation(entityId, entityType, version)).thenThrow(new NotFoundException());
```
Comment (john-hill) [inline services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/doi/DoiManagerImplTest.java:125]: remove test.

File: services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/doi/DoiManagerImplTest.java [hunk 9]
```diff
@@ -0,0 +1,567 @@
...
+		// Call under test
+		doiManager.getDoiAssociation( 12345L, null, entityType, version);
+	}
+
+	@Test(expected = IllegalArgumentException.class)
+	public void testGetAssociationNoObjectType() {
+		// Call under test
+		doiManager.getDoiAssociation( 12345L, entityId, null, version);
+	}
+
```
Comment (john-hill) [inline services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/doi/DoiManagerImplTest.java:147]: need a test for null version (should not throw an exception)

File: services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/doi/DoiManagerImplTest.java [hunk 10]
```diff
@@ -0,0 +1,567 @@
...
+		assertEquals(actualResponse.getCreators(), metadata.getCreators());
+		assertEquals(actualResponse.getTitles(), metadata.getTitles());
+		assertEquals(actualResponse.getPublicationYear(), metadata.getPublicationYear());
+		assertEquals(actualResponse.getResourceType(), metadata.getResourceType());
+		assertEquals(actualResponse.getStatus(), metadata.getStatus());
+	}
+
+	@Test(expected = UnauthorizedException.class)
+	public void testGetDoiUnauthorized() throws Exception {
+		when(mockAuthorizationManager.canAccess(any(UserInfo.class), any(String.class), any(ObjectType.class), any(ACCESS_TYPE.class)))
```
Comment (john-hill) [inline services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/doi/DoiManagerImplTest.java:171]: more targeted.

File: services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/doi/DoiManagerImplTest.java [hunk 11]
```diff
@@ -0,0 +1,567 @@
...
+		// Call under test
+		doiManager.getDoi( 12345L, entityId, entityType, version);
+	}
+
+	@Test
+	public void testCreateSuccess() throws Exception {
+		Doi dto = setUpDto(true);
+		when(mockDoiDao.getEtagForUpdate(dto.getObjectId(), dto.getObjectType(), dto.getObjectVersion())).thenThrow(new NotFoundException());
+		when(mockDoiDao.createDoiAssociation(dto)).thenReturn(dto);
+		doiManager.createOrUpdateDoi(12345L, dto);
```
Comment (john-hill) [inline services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/doi/DoiManagerImplTest.java:190]: call under test

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/doi/DoiManagerImpl.java [hunk 12]
```diff
@@ -0,0 +1,222 @@
...
+		// Check if the DOI exists, and if so, check to make sure eTags match.
+		try {
+			if (!doiAssociationDao.getEtagForUpdate(dto.getObjectId(), dto.getObjectType(), dto.getObjectVersion()).equals(dto.getEtag())) {
+				throw new ConflictingUpdateException("eTags do not match.");
+			}
+			doiAssociationDao.updateDoiAssociation(dto);
+		} catch (NotFoundException e1) {
+			try {
+				doiAssociationDao.createDoiAssociation(dto); //retry on fail
+			} catch (Exception e2) {
```
Comment (john-hill) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/doi/DoiManagerImpl.java:106]: catch the exception that is thrown for the case where duplicate key.

File: services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/doi/DoiManagerImplTest.java [hunk 13]
```diff
@@ -0,0 +1,567 @@
...
+		when(mockDoiDao.getDoiAssociation(entityId, entityType, version)).thenReturn(association);
+		doThrow(new ServiceUnavailableException()).when(mockDataciteClient).deactivate(any(String.class));
+		// Call under test
+		doiManager.deactivateDoi(1234L, entityId, entityType, version);
+		verify(mockDataciteClient, times(1)).deactivate(any(String.class));
+	}
+
+	@Test(expected = UnauthorizedException.class)
+	public void testDeactivateUnauthorized() throws Exception {
+		when(mockAuthorizationManager.canAccess(any(UserInfo.class), any(String.class), any(ObjectType.class), any(ACCESS_TYPE.class)))
```
Comment (john-hill) [inline services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/doi/DoiManagerImplTest.java:268]: more targeted

File: services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/doi/DoiManagerImplTest.java [hunk 14]
```diff
@@ -0,0 +1,567 @@
...
+	 * Test the expected race conditions
+	 */
+
+	/*
+	 * Test the case where an object is created between the client checking for an existing object and actually creating
+	 * that object. The client should retrieve an error guiding a retry. Upon retry, the object no longer exists and the
+	 * client succeeds.
+	 */
+	@Test
+	public void testDBFailedCreateWithSuccessOnRetry() throws Exception {
```
Comment (john-hill) [inline services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/doi/DoiManagerImplTest.java:360]: one test per branch through the code.

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/doi/DoiManagerImpl.java [hunk 15]
```diff
@@ -0,0 +1,222 @@
...
+		}
+
+		// Ensure the user is authorized to update the object that we are minting a DOI for
+		UserInfo currentUser = userManager.getUserInfo(userId);
+		UserInfo.validateUserInfo(currentUser);
+		AuthorizationManagerUtil.checkAuthorizationAndThrowException(
+				authorizationManager.canAccess(currentUser, dto.getObjectId(), dto.getObjectType(), ACCESS_TYPE.UPDATE));
+
+		// Check if the DOI exists, and if so, check to make sure eTags match.
+		try {
```
Comment (john-hill) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/doi/DoiManagerImpl.java:98]: maybe split into a new method createOrUpdateAssociation()

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/doi/DoiManagerImpl.java [hunk 16]
```diff
@@ -0,0 +1,222 @@
...
+			doiAssociationDao.updateDoiAssociation(dto);
+		} catch (NotFoundException e1) {
+			try {
+				doiAssociationDao.createDoiAssociation(dto); //retry on fail
+			} catch (Exception e2) {
+				throw new RecoverableMessageException(e2);
+			}
+		}
+
+		String uri = generateDoiUri(dto.getObjectId(), dto.getObjectType(), dto.getObjectVersion());
```
Comment (john-hill) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/doi/DoiManagerImpl.java:111]: split into createOrUpdateDataCite()
